api/api.go

```diff
@@ -278,46 +305,32 @@ func (api *API) listAlerts(w http.ResponseWriter, r *http.Request) {
 		}
 	}
 
-	alerts := api.alerts.GetPending()
-	defer alerts.Close()
-
-	// TODO(fabxc): enforce a sensible timeout.
-	for a := range alerts.Next() {
-		if err = alerts.Err(); err != nil {
-			break
-		}
-
-		if !alertMatchesFilterLabels(&a.Alert, matchers) {
-			continue
-		}
-
-		// Continue if alert is resolved
-		if !a.Alert.EndsAt.IsZero() && a.Alert.EndsAt.Before(time.Now()) {
-			continue
+	if receiverParam := r.FormValue("receiver"); receiverParam != "" {
+		re, err = regexp.Compile("^(?:" + receiverParam + ")$")
+		if err != nil {
+			respondError(w, apiError{
+				typ: errorBadData,
+				err: fmt.Errorf(
+					"failed to parse receiver param: %s",
+					receiverParam,
+				),
+			}, nil)
+			return
 		}
+	}
 
-		status := api.getAlertStatus(a.Fingerprint())
-
-		if !showSilenced && len(status.SilencedBy) != 0 {
-			continue
-		}
+	respond(w, flattenGroups(api.groups(matchers, showSilenced, re)))
+}
 
-		apiAlert := &APIAlert{
-			Alert:  &a.Alert,
-			Status: status,
+func flattenGroups(groups dispatch.AlertOverview) []*dispatch.APIAlert {
```

Comment:
Doesn't this result in duplicate alerts, because one alert can be present in multiple groups?
//CC @fabxc as I think we have talked about this before.